world-builder/world-gen/main.js

```diff
@@ -0,0 +1,276 @@
+        }
+
+        x = 158
+        for (let y = 98; y < 158; y++) {
+            let pixelVal = outArray[y][x];
+            if (pixelVal < minPixelAroundFlatSpot) {
+                minPixelAroundFlatSpot = pixelVal
+            }
+        }
+
+        console.log(minPixelAroundFlatSpot);
+
+        // Set the flat spot to the lowest surrounding terrain height.
+        for (let y = 99; y < 158; y++) {
+            for (let x = 99; x < 158; x++) {
+                n = minPixelAroundFlatSpot;
+                outArray[y][x] = n
+                let rgb = Math.round(255 * n);
+                ctx.fillStyle = "rgba(" + rgb + "," + rgb + "," + rgb + ",1.0)";
+                ctx.fillRect(x, y, 1, 1);
+            }
+        }
+
+        let flattenRange = 10
+
+        data = canv.getContext("2d").getImageData(0, 0, WIDTH, HEIGHT).data;
+
+        // Generate the actual terrain.
+        var vertices = plane.attributes.position.array;
+        for (i = 0, j = 2; i < data.length; i += 4, j += 3) {
+            vertices[j] = data[i] * HEIGHT_AMPLIFIER;
+        }
+
+        var mesh = new THREE.Mesh(plane, material);
+
+        plane.computeFaceNormals();
+        plane.computeVertexNormals();
+
+        GLOBALMIN = minPixelAroundFlatSpot
+        console.log(plane);
+        this.geometry = plane;
+    }
+});
+
+// Set the position of the terrain so it is on playable height.
+AFRAME.registerComponent('set-p', {
+    dependencies: ['position'],
+  
+    init: function () {
+        pos = {'x': 0, 'y': -(GLOBALMIN * 5 * 10)-.2, 'z': 0}
+        this.el.setAttribute('position', pos);
+    }
+  });
+
+// TODO: REPLACE WITH THREE JS VECTOR?
+// Very simple class to create a vector
+class Vector2 {
+    constructor(x, y) {
+        this.x = x;
+        this.y = y;
+    }
+    dot(other) {
+        return this.x * other.x + this.y * other.y;
+    }
+}
+
+// Shuffle the permutation table.
+function shuffle(tab) {
+    for (let e = tab.length - 1; e > 0; e--) {
+        let index = Math.round(Math.random() * (e - 1)),
+            temp = tab[e];
+
+        tab[e] = tab[index];
+        tab[index] = temp;
+    }
+}
+
+// Make the permutation array.
+function makePermutation() {
+    let P = [];
+    for (let i = 0; i < 256; i++) {
+        P.push(i);
+    }
+    shuffle(P);
+    for (let i = 0; i < 256; i++) {
+        P.push(P[i]);
+    }
+
+    return P;
+}
+let P = makePermutation();
+
+// Contant vector for each value.
+function getConstantVector(v) {
+    //v is the value from the permutation table.
+    let h = v & 3;
+    if (h == 0)
+        return new Vector2(1.0, 1.0);
+    else if (h == 1)
+        return new Vector2(-1.0, 1.0);
+    else if (h == 2)
+        return new Vector2(-1.0, -1.0);
+    else
+        return new Vector2(1.0, -1.0);
+}
+
+function fade(t) {
+    return ((6 * t - 15) * t + 10) * t * t * t;
+}
+
+function lerp(t, a1, a2) {
+    return a1 + t * (a2 - a1);
+}
+
+// Puts all the function together.
+function noise2D(x, y) {
```

Comment:
A more elaborate comment/docstring on this function and maybe the others would help with cod eclarification